python/cudf/cudf/core/join/join.py

```diff
@@ -0,0 +1,454 @@
+        index_dtypes = []
+        data_dtypes = {}
+        for name, col in itertools.chain(
+            self.lhs._data.items(), self.rhs._data.items()
+        ):
+            data_dtypes[name] = col.dtype
+
+        if self.left_index and self.right_index:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs.index._data.items(), self.rhs.index._data.items()
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(
+                    vl, vr, self.how
+                )
+                index_dtypes.append(to_dtype)
+        if self.left_on and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs[self.left_on]._data.items(),
+                self.rhs[self.right_on]._data.items(),
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(
+                    vl, vr, self.how
+                )
+                data_dtypes[kl] = to_dtype
+                data_dtypes[kr] = to_dtype
+        if self.left_index and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs.index._data.items(),
+                self.rhs[self.right_on]._data.items(),
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(
+                    vl, vr, self.how
+                )
+                data_dtypes[kr] = to_dtype
+                index_dtypes = [
+                    col.dtype for col in self.rhs.index._data.values()
+                ]
+        if self.right_index and self.left_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs[self.left_on]._data.items(),
+                self.rhs.index._data.items(),
+            ):
+                to_dtype = self.libcudf_to_output_casting_rules(
+                    vl, vr, self.how
+                )
+                data_dtypes[kl] = to_dtype
+                index_dtypes = [
+                    col.dtype for col in self.lhs.index._data.values()
+                ]
+        return (index_dtypes, data_dtypes)
+
+    def typecast_libcudf_to_output(self, output, output_dtypes):
+        index_dtypes, data_dtypes = output_dtypes
+        if output._index and len(index_dtypes) > 0:
+            for i, (index_col_lbl, index_col) in enumerate(
+                output._index._data.items()
+            ):
+                index_dtype = index_dtypes[i]
+                if index_dtype:
+                    output._index._data[
+                        index_col_lbl
+                    ] = self._build_output_col(index_col, index_dtype)
+        for data_col_lbl, data_col in output._data.items():
+            data_dtype = data_dtypes[data_col_lbl]
+            if data_dtype:
+                output._data[data_col_lbl] = self._build_output_col(
+                    data_col, data_dtype
+                )
+        return output
+
+    def _build_output_col(self, col, dtype):
+        if is_categorical_dtype(dtype) and dtype != "category":
```

Comment:
When does this return True? If dtype is a `cudf.CategoricalDtype` or a `pd.CategoricalDtype` then `== "category"` would return True.

Reply:
This wasn't working correctly and has been updated, but the case it was supposed to handle still needs to be handled. This is encountered when of the original columns of one of the operands needs to be cast back to categorical after the join is complete, since it's processed through the gather as a column of ints. Since this process might change what categories exist in the output, the objective here is to just construct a new categorical column without predetermined categories. Unless this is handled separately, we'll get an error when we try and access the `categories` attribute of the datatype as this is just a string.